File: Project3/server/src/main/java/bgu/spl/net/impl/stomp/StompFrame.java
package bgu.spl.net.impl.stomp;

import java.util.HashMap;
import java.util.Map;

public class StompFrame {

    private final String command;
    private final Map<String, String> headers;
    private final String body;

    public StompFrame(String command, String hds, String body) {
        this.command = command;
        this.headers = new HashMap<String, String>();
        headers.put("receipt-id", hds);
        this.body = body;
    }

    public StompFrame(String command, Map<String, String> headers, String body) {
        this.command = command;
        this.headers = headers;
        this.body = body;
    }

    public StompFrame(String msgError, int msgId) {
        this.command = "ERROR";
        this.headers = new HashMap<String, String>();
        headers.put("receipt-id", msgId + "");
        headers.put("message", msgError);
        this.body = "";
    }

    public StompFrame() {
        this.command = "";
        this.headers = new HashMap<String, String>();
        this.body = "";
    }

    public String getCommand() {
        return command;
    }

    public Map<String, String> getHeaders() {
        return headers;
    }

    public String getBody() {
        return body;
    }

    public String toString() {
        String res = command + "\n";
        for (String key : headers.keySet()) {
            res += key + ":" + headers.get(key) + "\n";
        }
        if (body != null)
            res += "\n" + body; 
        else
            res += "\n";

        return res;
    }

    public static String connect() {
        Map<String, String> headers = new HashMap<String, String>();
        headers.put("version", "1.2");
        return (new StompFrame("CONNECTED", headers, "")).toString();
    }

    public String message(int subId, int messageId, String destination, String msg) {
        String res = "MESSAGE\n" +
                    "subscription:" + subId + "\n" + 
                    "message-id:" + messageId + "\n" +
                    "destination:" + destination + "\n\n" +
                    msg;
        return res;
    }

    // public static void main(String[] args) {
    //     HashMap<String, String> headers = new HashMap<String, String>();
    //     headers.put("accept-version", "1.2");
    //     headers.put("host", "stomp.cs.bgu.ac.il");
    //     headers.put("login", "fawzi");
    //     headers.put("passcode", "films");
    //     String stri = new StompFrame("CONNECT", headers, "").toString();
    //     String[] str = stri.split("\n");
    //     String username = str[3].split(":")[1]; // get username
    //     String password = str[4].split(":")[1]; // get pw

    // }

}

File: Project3/client/src/StompFrame.cpp

#include "../include/StompFrame.h"
#include <map>
//#include "../src/event.cpp"
//#include <map>

StompFrame::StompFrame() : command(""), body(""), headers(){};

 std::string StompFrame::login(string username, string password)
{
    command = "CONNECT";
    headers["accept-version"] = "1.2";
    headers["host"] = "stomp.cs.bgu.ac.il";
    headers["login"] = username;
    headers["passcode"] = password;
    // body is empty.
    return this->toString();
}
 std::string StompFrame::logout(int receipt)
{
    command = "DISCONNECT";
    headers["receipt"] = std::to_string(receipt);

    // body is empty.
    return this->toString();
}

std::string StompFrame::join(string destination, int subId, int receipt)
{
    command = "SUBSCRIBE";
    headers["destination"] = "/" + destination;
    headers["id"] = std::to_string(subId);
    headers["receipt"] = std::to_string(receipt);
    // body is empty.
    return this->toString();
}

std::string StompFrame::exit(int subId, int receipt)
{
    command = "UNSUBSCRIBE";
    headers["id"] = std::to_string(subId);
    headers["receipt"] = std::to_string(receipt);
    // body is empty.
    return this->toString();
}
std::string getStringUpdate(std::map<string,string> updatesMap){
    string res;
    for (auto it = updatesMap.begin(); it != updatesMap.end(); it++) {
        // Get the key and value of the current element
        std::string key = it->first;
        std::string value = it->second;
        res += "\t" + key + ": " + value + "\n";
        // Print the key and value
    }
    return res;
 }
std::string StompFrame::report(string username, string destination,  Event file)
{
     command = "SEND";
     headers["destination"] = "/" + destination;
     std::string str[] = {"user: " + username,
                           "team a: " +  file.get_team_a_name(),
                           "team b: " +file.get_team_b_name(),
                           "event name: " + file.get_name(),
                           "time: " + std::to_string(file.get_time()) + "\n",
                           "general game updates: \n" + getStringUpdate(file.get_game_updates()),
                           "team a updates: \n"  + getStringUpdate(file.get_team_a_updates()),
                           "team b updates: \n" + getStringUpdate(file.get_team_b_updates()),
                           "description: " + file.get_discription()};
    bool jumpNow = false;
    for(string s: str){
        if(split(s,':')[0] == "time")
            jumpNow = true;
        if(jumpNow)
            body += s ;
        else
            body += s + "\n";
    }
    return this->toString();
}
std::string StompFrame::toString() {
    std::string frameSent = command + "\n";
    for (auto it = headers.begin(); it != headers.end(); it++) {
        // Get the key and value of the current element
        std::string key = it->first;
        std::string value = it->second;
        frameSent += key + ":" + value + "\n";
        // Print the key and value
    }

    bool jumpNow = false;
    if (!body.empty()) {
        frameSent += "\n" + body;
    }
//    frameSent += "\n" + body;
    return frameSent;
}

 std::vector<std::string> StompFrame::split(const std::string &str, char delimiter)
 {
     std::vector<std::string> tokens;
     std::string token;
     std::size_t start = 0, end = 0;
     while ((start = str.find_first_not_of(delimiter, end)) != std::string::npos)
     {
         end = str.find_first_of(delimiter, start);
         token = str.substr(start, end - start);
         tokens.push_back(token);
     }
     return tokens;
 }
